File: src/main/java/com/matamercer/microblog/security/jwt/JwtUtil.java
package com.matamercer.microblog.security.jwt;

import com.matamercer.microblog.models.entities.RefreshToken;
import com.matamercer.microblog.models.entities.User;
import com.matamercer.microblog.models.repositories.RefreshTokenRepository;
import com.matamercer.microblog.models.repositories.UserRepository;
import com.matamercer.microblog.security.UserRole;
import com.matamercer.microblog.web.error.AuthenticationException;
import com.matamercer.microblog.web.error.UserNotFoundException;
import io.jsonwebtoken.*;
import lombok.var;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.crypto.SecretKey;
import javax.servlet.http.HttpServletResponse;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

@Service
public class JwtUtil {

    private final SecretKey secretKey;
    private final JwtConfig jwtConfig;
    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;

    @Autowired
    public JwtUtil(SecretKey secretKey, JwtConfig jwtConfig, UserRepository userRepository, RefreshTokenRepository refreshTokenRepository) {
        this.secretKey = secretKey;
        this.jwtConfig = jwtConfig;
        this.userRepository = userRepository;
        this.refreshTokenRepository = refreshTokenRepository;
    }



    @Transactional
    public String createAccessToken(Long userId){
        var optionalUser = userRepository.findById(userId);
        return createAccessTokenHelper(optionalUser);
    }

    @Transactional
    public String createAccessToken(String username){
        var optionalUser = userRepository.findByUsername(username);
        return createAccessTokenHelper(optionalUser);
    }

    private String createAccessTokenHelper(Optional<User> optionalUser){
        if(optionalUser.isPresent()) {
            User user = optionalUser.get();
            return createToken(getUserClaims(user.getId(), user.getUsername(), user.getRole()), addHoursToCurrentDate(jwtConfig.getAccessTokenExpirationInHours()));
        }else{
            throw new UserNotFoundException("Error creating access token. User not found.");
        }
    }

    @Transactional
    public String createRefreshToken(Long userId){
        var optionalUser = userRepository.findById(userId);
        return createRefreshTokenHelper(optionalUser);
    }

    @Transactional
    public String createRefreshToken(String username){
        var optionalUser = userRepository.findByUsername(username);
        return createRefreshTokenHelper(optionalUser);
    }

    private String createToken(Map<String, ?> claims, Date exp){
        return Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(new Date())
                .setExpiration(exp)
                .signWith(secretKey)
                .compact();
    }

    private String createRefreshTokenHelper(Optional<User> optionalUser){
        if(optionalUser.isPresent()){
            User user = optionalUser.get();
            RefreshToken refreshTokenEntity = refreshTokenRepository.save(new RefreshToken(user));
            var claims = getUserClaims(user.getId(), user.getUsername(), user.getRole());
            claims.put("refreshTokenEntityId", refreshTokenEntity.getId().toString());
            return createToken(claims, java.sql.Date.valueOf(LocalDate.now().plusDays(jwtConfig.getRefreshTokenExpirationInDays())));
        }else{
            throw new UserNotFoundException("Error creating refresh token. User not found.");
        }
    }

    private Map<String, String> getUserClaims(Long userId, String username, UserRole userRole ){
        Map<String, String> claims = new HashMap<>();
        claims.put("userId", userId.toString());
        claims.put("username", username);
        claims.put("userRole", userRole.toString());
        return claims;
    }

    private Date addHoursToCurrentDate(int hours){
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(new Date());
        calendar.add(Calendar.HOUR_OF_DAY, hours);
        return calendar.getTime();
    }

    public Claims extractAllClaims(String token){
        try {
            return Jwts
                    .parserBuilder().setSigningKey(secretKey)
                    .build()
                    .parseClaimsJws(token).getBody();
        } catch(ExpiredJwtException expiredJwtException) {
            throw new AuthenticationException("Token is expired", expiredJwtException);
        } catch(UnsupportedJwtException unsupportedJwtException) {
            throw new AuthenticationException("This JWT token is unsupported", unsupportedJwtException);
        } catch(MalformedJwtException malformedJwtException) {
            throw new AuthenticationException("Malformed JWT", malformedJwtException);
        } catch(SignatureException signatureException) {
            throw new AuthenticationException("JWT Signature invalid", signatureException);
        } catch(IllegalArgumentException illegalArgumentException) {
            throw new AuthenticationException("Illegal argument", illegalArgumentException);
        }
    }

    public boolean isTokenExpired(Claims claims){
        return claims.getExpiration().before(new Date());
    }

    public Long getUserId(Claims claims){
        return Long.parseLong(claims.get("userId").toString());
    }

    public UserRole getUserRole(Claims claims){
        return UserRole.valueOf(claims.get("userRole").toString());
    }

    public String getUserName(Claims claims){
        return claims.get("username").toString();
    }






}

File: src/main/java/com/matamercer/microblog/security/authentication/JwtUtil.kt
package com.matamercer.microblog.security.authentication

import com.matamercer.microblog.models.entities.RefreshToken
import com.matamercer.microblog.models.entities.User
import com.matamercer.microblog.models.repositories.RefreshTokenRepository
import com.matamercer.microblog.models.repositories.UserRepository
import com.matamercer.microblog.security.authorization.UserRole
import com.matamercer.microblog.web.error.exceptions.AuthenticationException
import com.matamercer.microblog.web.error.exceptions.UserNotFoundException
import io.jsonwebtoken.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.sql.Date
import java.time.LocalDate
import java.util.*
import javax.crypto.SecretKey
import javax.transaction.Transactional

@Service
class JwtUtil @Autowired constructor(
    private val secretKey: SecretKey,
    private val jwtConfig: JwtConfig,
    private val userRepository: UserRepository,
    private val refreshTokenRepository: RefreshTokenRepository
) {
    @Transactional
    fun createAccessToken(userId: Long): String {
        val user = userRepository.findById(userId).get()
        return createAccessTokenHelper(user)
    }

    @Transactional
    fun createAccessToken(username: String?): String {
        val user = userRepository.findByUsername(username)
        return createAccessTokenHelper(user!!)
    }

    private fun createAccessTokenHelper(user: User): String {
        return jwtConfig.tokenPrefix + createToken(
            getUserClaims(user.id, user.username, user.role, user.activeBlog?.id),
            addHoursToCurrentDate(jwtConfig.accessTokenExpirationInHours)
        )
    }

    @Transactional
    fun createRefreshToken(userId: Long): String {
        val user = userRepository.findById(userId).get()
        return createRefreshTokenHelper(user)
    }

    @Transactional
    fun createRefreshToken(username: String?): String {
        val user = userRepository.findByUsername(username)
        return createRefreshTokenHelper(user!!)
    }

    private fun createToken(claims: Map<String, String>, exp: java.util.Date): String {
        return Jwts.builder()
            .setClaims(claims)
            .setIssuedAt(Date())
            .setExpiration(exp)
            .signWith(secretKey)
            .compact()
    }

    private fun createRefreshTokenHelper(user: User): String {
        val refreshTokenEntity = refreshTokenRepository.save(RefreshToken(user))
        val claims = getUserClaims(user.id, user.username, user.role, user.activeBlog?.id)
        claims["refreshTokenEntityId"] = refreshTokenEntity.id.toString()
        return createToken(
            claims,
            Date.valueOf(LocalDate.now().plusDays(jwtConfig.refreshTokenExpirationInDays.toLong()))
        )
    }

    private fun getUserClaims(
        userId: Long?,
        username: String?,
        userRole: UserRole?,
        activeBlogId: Long?
    ): MutableMap<String, String> {
        val claims: MutableMap<String, String> = HashMap()
        claims["userId"] = userId.toString()
        claims["username"] = username.toString()
        claims["userRole"] = userRole.toString()
        claims["activeBlogId"] = activeBlogId.toString()
        return claims
    }

    private fun addHoursToCurrentDate(hours: Int): java.util.Date {
        val calendar = Calendar.getInstance()
        calendar.time = Date()
        calendar.add(Calendar.SECOND, hours)
        return calendar.time
    }

    fun extractAllClaims(token: String?): Claims {
        return try {
            Jwts
                .parserBuilder().setSigningKey(secretKey)
                .build()
                .parseClaimsJws(token).body
        } catch (expiredJwtException: ExpiredJwtException) {
            throw AuthenticationException("Token is expired", expiredJwtException)
        } catch (unsupportedJwtException: UnsupportedJwtException) {
            throw AuthenticationException("This JWT token is unsupported", unsupportedJwtException)
        } catch (malformedJwtException: MalformedJwtException) {
            throw AuthenticationException("Malformed JWT", malformedJwtException)
        } catch (signatureException: SignatureException) {
            throw AuthenticationException("JWT Signature invalid", signatureException)
        } catch (illegalArgumentException: IllegalArgumentException) {
            throw AuthenticationException("Illegal argument", illegalArgumentException)
        }
    }

    fun isTokenExpired(claims: Claims): Boolean {
        return claims.expiration.before(Date())
    }

    fun getUserId(claims: Claims): Long {
        return claims["userId"].toString().toLong()
    }

    fun getUserRole(claims: Claims): UserRole {
        return UserRole.valueOf(claims["userRole"].toString())
    }

    fun getUserName(claims: Claims): String {
        return claims["username"].toString()
    }
}